specs/waku/v2/waku-v2-rpc-api.md

```diff
@@ -0,0 +1,459 @@
+
+#### Response
+
+- **`Bool`** - `true` on success or an [error](https://www.jsonrpc.org/specification#error_object) on failure.
+
+# Example usage
+
+## Store API
+
+### `get_waku_v2_store_v1_messages`
+
+This method is part of the `store` API and the specific resources to retrieve are (historical) `messages`. The protocol (`waku`) is on `v2`, whereas the Store API definition is on `v1`.
+
+1. `get` *all* the historical messages for content topic **1**; no paging required
+
+#### Request
+
+```curl -d '{"jsonrpc":"2.0","id":"id","method":"get_waku_v2_store_v1_messages", "params":[[1]]}' --header "Content-Type: application/json" http://localhost:8545```
+
+```jsonrpc
+{
+  "jsonrpc": "2.0",
+  "id": "id",
+  "method": "get_waku_v2_store_v1_messages",
+  "params": [
+    [
+      1
+    ]
+  ]
+}
+```
+
+#### Response
+
+```jsonrpc
+{
+  "jsonrpc": "2.0",
+  "id": "id",
+  "result": {
+    "messages": [
+      {
+        "payload": [
+          1
+        ],
+        "contentTopic": 1,
+        "version": 0
+      },
+      {
+        "payload": [
+          2
+        ],
+        "contentTopic": 1,
+        "version": 0
+      },
+      {
+        "payload": [
+          3
+        ],
+        "contentTopic": 1,
+        "version": 0
+      }
+    ],
+    "pagingInfo": null
+  },
+  "error": null
+}
+```
+
+---
+
+2. `get` a single page of historical messages for content topic **1**; 2 messages per page, backward direction. Since this is the initial query, no `cursor` is provided, so paging will be performed from the end of the list.
+
+#### Request
+
+```curl -d '{"jsonrpc":"2.0","id":"id","method":"get_waku_v2_store_v1_messages", "params":[[1],{"pageSize":2,"forward":false}]}' --header "Content-Type: application/json" http://localhost:8545```
+
+```jsonrpc
+{
+  "jsonrpc": "2.0",
+  "id": "id",
+  "method": "get_waku_v2_store_v1_messages",
+  "params": [
+    [
+      1
+    ],
+    {
+      "pageSize": 2,
+      "forward": false
+    }
+  ]
+}
+```
+
+#### Response
+
+```jsonrpc
+{
+  "jsonrpc": "2.0",
+  "id": "id",
+  "result": {
+    "messages": [
+      {
+        "payload": [
+          2
+        ],
+        "contentTopic": 1,
+        "version": 0
+      },
+      {
+        "payload": [
+          3
+        ],
+        "contentTopic": 1,
+        "version": 0
+      }
+    ],
+    "pagingInfo": {
+      "pageSize": 2,
+      "cursor": {
+        "digest": "abcdef",
+        "receivedTime": 1605887187
+      },
+      "forward": false
+    }
+  },
+  "error": null
+}
+```
+
+---
+
+3. `get` the next page of historical messages for content topic **1**, using the cursor received above; 2 messages per page, backward direction.
+
+#### Request
+
+```curl -d '{"jsonrpc":"2.0","id":"id","method":"get_waku_v2_store_v1_messages", "params":[[1],{"pageSize":2,"cursor":{"digest":"abcdef","receivedTime":1605887187.00},"forward":false}]}' --header "Content-Type: application/json" http://localhost:8545```
```

Comment:
I'm a bit confused here, it was my understanding that we can't use named parameters for content topic, but here we are doing that for pageSize through a map. Is there a reason we can't do the same for content topic? A la `{"contentTopic" 1}` or `{"contentTopics" [1,2,3]}`

Reply:
The `params` itself are unnamed, positional and in an array. The fields within an `Object` param _can_ be named, though the `Object` they're a part of will remain unnamed.
`pageSize`, etc. are only named then because they're fields in an (unnamed) param `Object` called `pagingOptions`.
We could hack everything to have names by wrapping all `params` in a `Request` envelope `Object`, though this will still have to be nested within the `params` `Array` and require such an envelope for all methods.